[C10D] Add time_created_us to flight recorder

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114817
* #114615
* #114625
* __->__ #114810

time_created_us is the cpu-side epoch_time (in usec) when a flight-recorder
event was created. It loosely corresponds to the time the c10d collective
API was called and a work object was created.  It does NOT correspond to
the time the collective started on the GPU.

We follow the precedent of us epoch time from this PR adding timestamps
to the cuda caching allocator:
https://github.com/pytorch/pytorch/pull/112266

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @kiukchung @d4l3k @lucasllc

wconstab committed:
[C10D] Add time_created_us to flight recorder

time_created_us is the cpu-side epoch_time (in usec) when a flight-recorder
event was created. It loosely corresponds to the time the c10d collective
API was called and a work object was created.  It does NOT correspond to
the time the collective started on the GPU.

We follow the precedent of us epoch time from this PR adding timestamps
to the cuda caching allocator:
https://github.com/pytorch/pytorch/pull/112266

[ghstack-poisoned]

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -492,6 +503,7 @@ struct NCCLTraceBuffer {
       dict.insert(pg_id_s, int64_t(e.pg_id_));
       dict.insert(seq_id_s, int64_t(e.seq_id_));
       dict.insert(profiling_name_s, e.profiling_name_);
+      dict.insert(time_created_s, int64_t(e.time_created_));
```
I think time conversion goes here.

zdevito commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -374,14 +380,18 @@ struct NCCLTraceBuffer {
         std::move(start),
-        std::move(end)};
+        std::move(end),
+        uint64_t(tsc_to_ns(c10::getApproximateTime()) / 1000)};
```
and we store just getApproximateTime here.

wconstab replied:
yep, agreed.

wconstab committed:
Update on "[C10D] Add time_created_us to flight recorder"


time_created_us is the cpu-side epoch_time (in usec) when a flight-recorder
event was created. It loosely corresponds to the time the c10d collective
API was called and a work object was created.  It does NOT correspond to
the time the collective started on the GPU.

We follow the precedent of us epoch time from this PR adding timestamps
to the cuda caching allocator:
https://github.com/pytorch/pytorch/pull/112266

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Add time_created_us to flight recorder"


time_created_us is the cpu-side epoch_time (in usec) when a flight-recorder
event was created. It loosely corresponds to the time the c10d collective
API was called and a work object was created.  It does NOT correspond to
the time the collective started on the GPU.

We follow the precedent of us epoch time from this PR adding timestamps
to the cuda caching allocator:
https://github.com/pytorch/pytorch/pull/112266

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
